source/common/protobuf/utility.cc

```diff
@@ -375,42 +376,25 @@ void checkForDeprecatedNonRepeatedEnumValue(const Protobuf::Message& message,
   }
 }
 
-void checkForUnexpectedFields(const Protobuf::Message& message,
-                              ProtobufMessage::ValidationVisitor& validation_visitor,
-                              Runtime::Loader* runtime) {
-  // Reject unknown fields.
-  const auto& unknown_fields = message.GetReflection()->GetUnknownFields(message);
-  if (!unknown_fields.empty()) {
-    std::string error_msg;
-    for (int n = 0; n < unknown_fields.field_count(); ++n) {
-      if (unknown_fields.field(n).number() == ProtobufWellKnown::OriginalTypeFieldNumber) {
-        continue;
-      }
-      error_msg += absl::StrCat(n > 0 ? ", " : "", unknown_fields.field(n).number());
-    }
-    // We use the validation visitor but have hard coded behavior below for deprecated fields.
-    // TODO(htuch): Unify the deprecated and unknown visitor handling behind the validation
-    // visitor pattern. https://github.com/envoyproxy/envoy/issues/8092.
-    if (!error_msg.empty()) {
-      validation_visitor.onUnknownField("type " + message.GetTypeName() +
-                                        " with unknown field set {" + error_msg + "}");
-    }
-  }
+class UnexpectedFieldProtoVisitor : public ProtobufMessage::ConstProtoVisitor {
+public:
+  UnexpectedFieldProtoVisitor(ProtobufMessage::ValidationVisitor& validation_visitor,
+                              Runtime::Loader* runtime)
+      : validation_visitor_(validation_visitor), runtime_(runtime) {}
 
-  const Protobuf::Descriptor* descriptor = message.GetDescriptor();
-  const Protobuf::Reflection* reflection = message.GetReflection();
-  for (int i = 0; i < descriptor->field_count(); ++i) {
-    const Protobuf::FieldDescriptor* field = descriptor->field(i);
-    absl::string_view filename = filenameFromPath(field->file()->name());
+  const void* onField(const Protobuf::Message& message, const Protobuf::FieldDescriptor& field,
+                      const void*) override {
+    const Protobuf::Reflection* reflection = message.GetReflection();
+    absl::string_view filename = filenameFromPath(field.file()->name());
 
     // Before we check to see if the field is in use, see if there's a
     // deprecated default enum value.
-    checkForDeprecatedNonRepeatedEnumValue(message, filename, field, reflection, runtime);
+    checkForDeprecatedNonRepeatedEnumValue(message, filename, &field, reflection, runtime_);
 
     // If this field is not in use, continue.
-    if ((field->is_repeated() && reflection->FieldSize(message, field) == 0) ||
-        (!field->is_repeated() && !reflection->HasField(message, field))) {
-      continue;
+    if ((field.is_repeated() && reflection->FieldSize(message, &field) == 0) ||
+        (!field.is_repeated() && !reflection->HasField(message, &field))) {
```

Comment:
Surprisingly tricky, since what is counted as used/unused is subtle for scalar fields. This can be seen in this code, since it first performs the scalar enum check before conditioning on whether the field is set in some sense.